### PR TITLE
WebAC ACL Interface

### PR DESCRIPTION
**Pull request by whikloj**

### CODE REVIEW

File: src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.fcrepo.kernel.api.models.FedoraResource;
+
+import com.hp.hpl.jena.vocabulary.RDF;
+
+/**
+ * Interface for WebAC ACL objects.
+ * @author whikloj
+ * @since 2015-08-25
+ */
+public interface AbstractFedoraWebAcl {
```
Comment (awoods) [inline src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java]: Why "Abstract"FedoraWebAcl? Any reason not to drop the "Abstract" prefix?

File: src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return set of acl:agentClass URIs
+     */
+    Set<URI> getGroups();
+
+    /*
+     * Set/Modify the acl:modes of this ACL.
+     * @param users a Set of permissions
+     */
+    void setPermissions(Set<URI> permissions);
```
Comment (awoods) [inline src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java]: Maybe we should use the webAC terminology of "modes" vs. "permissions".

Comment (peichman-umd) [inline src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java]: I agree with @awoods on this.

File: src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java (hunk 1 above)
Comment (ajs6f) [inline src/main/java/org/fcrepo/auth/webac/AbstractFedoraWebAcl.java]: This is an interface. It is in the Fedora codebase. Why does it have either `Abstract` _or_ `Fedora` as part of its name?

**Comment by peichman-umd** [comment]
Would this interface be able to support a method that takes a resource and/or an agent (as URIs) and returns the authorization(s) that mention that resource and/or agent? Alternatively, a method that returned the list of authorizations in the ACL, and the client class could iterate over that list to find the authorizations that were relevant to its interests? If either of these are desired, I would suggest an additional interface/class to represent a single authorization ("authorization" being the WebAC terminology for a single rule relating an agent(Class) to an accessTo(Class) resource with an access mode.

**Comment by awoods** [comment]
@peichman-umd, Although per the WebAC "spec", agents are represented as URIs. My understanding was that most installations would only be able to support Strings.

How are you defining "authorizations"? I thought this interface represented a single "authorization", i.e. all principals that have these modes on these resources (and resource classes).

**Comment by peichman-umd** [comment]
@awoods, The way I read the WebAC "spec", an ACL is a graph containing one or more acl:Authorization resources, which all seem to be represented as blank nodes with acl:accessTo, acl:agent, acl:mode, etc. properties. Thus I was assuming that an ACL class would represent that graph. 

Maybe this is a case of terminological confusion? If so, I would suggest renaming this interface to something like WebACAuthorization.

**Comment by peichman-umd** [comment]
@awoods, For the issue of representing agents as strings or URIs, would the new `acct:` URI scheme ([RFC 7565](https://tools.ietf.org/html/rfc7565)) help at all? It is specifically designed to make no commitment about how the URIs are resolved, it is just for naming accounts on a particular system.  The only issue might be what to put in the host section of the URI.

**Comment by awoods** [comment]
From: https://wiki.duraspace.org/display/FF/2015-07-22+-+WebAccessControl+Authorization+Delegate+Planning+Meeting
"Probably not using URIs for agents at least at first"

**Comment by ajs6f** [comment]
Those `acl:Authorization`s don't _have_ to be blank nodes, right? They could be something more sane, like hash-URI nodes.

**Comment by whikloj** [comment]
My grasp of some of the W3C lingo is weak. But based on this
"Note that there are properties of an Authorization which allow you to specify who gets access either by giving a relation to a specific user or a different relation to a class of user. The same technique is used for expressing to which resource or class of resources access is being granted."

My assumption is that one Authorization contains:
* 1 or more triples with a predicate from the set [acl:agent, acl:agentClass]
* 1 or more triples with a predicate from the set [acl:accessTo, acl:accessToClass]
* 1 triple with the predicate acl:mode

**Comment by awoods** [comment]
@whikloj, I had a similar representation in mind. I think we need to take a moment and decide on a clear, clean, consistent perspective on how Authorizations are represented in the repository.

**Comment by peichman-umd** [comment]
@ajs6f Yes, the Authorizations could definitely be hash-URI nodes.
@whikloj Yes, that fits with my reading of Authorization as well.

**Comment by whikloj** [comment]
Ok, I can change the name to WebACAuthorization. 

As for your first question @peichman-umd. 

As the acl:mode will be the same for all users and resources referenced by this ACL. 
The only check is whether the agent exists in the ACL. Not sure this is the right place for that check.

**Comment by awoods** [comment]
Is there a reason to even create hash resources? In other words, does an ACL resource need to contain a sub-graph of Authorizations, versus simply having all of the triples @whikloj described attached to the ACL resource itself?

**Comment by whikloj** [comment]
@awoods, I don't think so. Based on my understanding we would have simple triples on the Acl resource where the Acl resource would always be the subject.

**Comment by peichman-umd** [comment]
@whikloj: The way I have been reading the W3C spec is that the ACL is never actually the subject of any triples. There is an `acl:acl` property on a protected resource that identifies the location of its ACL, but that is just a URI to be retrieved.

**Comment by awoods** [comment]
@peichman-umd, are you saying that your understanding is that "protected resources" have a property that points to their ACL?

**Comment by peichman-umd** [comment]
@awoods That was my understanding of this section of the spec: http://www.w3.org/wiki/WebAccessControl#Modifying_Access_Control_information

**Comment by whikloj** [comment]
@peichman-umd, yes I see what you mean.
But then I can't understand how we handle the accessToClass functions. If the acl is not the subject of the triple that provides acl:accessToClass to rdf:type, what is?

**Comment by peichman-umd** [comment]
@whikloj, I think it is the Authorization. Note that the [ontology](http://www.w3.org/ns/auth/acl) does not actually define a class for the ACLs, which suggests that they are not intended to be RDF resources themselves. Instead, I think the ACL is supposed to be used as just a URL where a graph of authorizations can be retrieved.

**Comment by ajs6f** [comment]
I have no understanding of nor opinion about this issue, but just from the diction and tone of this conversation, perhaps this is a question best put to whomever is writing or maintaining WebAC?

**Comment by peichman-umd** [comment]
Also, from the [WAC and LDP](http://www.w3.org/wiki/WebAccessControl#WAC_And_LDP) section:
> So even though the acl file is neither an LDPR nor an LDPC it should be editable using the same methods.

**Comment by ajs6f** [comment]
This seems a bit confusing: maybe check with the (ex-)LDP WG for some clarification? @azaroth42?

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * @author whikloj
+ * @since 2015-08-25
+ */
+public interface WebACAuthorization {
+
+    /*
+     * Set/Modify the acl:agents of this ACL.
+     * @param users a Set of agents
+     */
+    void setUsers(Set<String> users);
```
Comment (awoods) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: Should we use `acl:Agent` terminology? If so, are userIds different than groupIds?

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return set of agents
+     */
+    Set<String> getUsers();
+
+    /*
+     * Set/Modify the acl:agentClass of this ACL.
+     * @param users a Set of agentClasses
+     */
+    void setGroups(Set<URI> groups);
```
Comment (awoods) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: I do not think our notion of "groups" is the same as `acl:agentClass`. Although the `Set<URI>` parameter to this method seems inline with the spec, the method name is confusing.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+ * @author whikloj
+ * @since 2015-08-25
+ */
+public interface WebACAuthorization {
+
+    /*
+     * Set/Modify the acl:agents of this ACL.
+     * @param agents a Set of acl:agents
+     */
+    void setAgents(Set<String> agents);
```
Comment (acoburn) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: I think we should avoid all setters in this interface. An `Authorization` should, effectively, be immutable. I'd like to see all properties set in the constructor of the implementing class.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return set of acl:agents
+     */
+    Set<String> getAgents();
+
+    /*
+     * Set/Modify the acl:agentClass of this ACL.
+     * @param agentClasses a Set of acl:agentClasses
+     */
+    void setAgentClasses(Set<URI> agentClasses);
```
Comment (acoburn) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: See comment above about no setters.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return set of acl:agentClass URIs
+     */
+    Set<URI> getAgentClasses();
+
+    /*
+     * Set/Modify the acl:modes of this ACL.
+     * @param modes a Set of acl:modes
+     */
+    void setModes(Set<URI> modes);
```
Comment (acoburn) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: See comment above about no setters.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return set of acl:modes
+     */
+    Set<URI> getModes();
+
+    /*
+     * Set/Modify the direct links targets of this ACL.
+     * @param resources a Set of FedoraResources
+     */
+    void setObjects(Set<FedoraResource> resources);
```
Comment (acoburn) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: See comment above about no setters.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return set of FedoraResource
+     */
+    Set<FedoraResource> getObjects();
+
+    /*
+     * Set/Modify the rdf:types to apply this ACL to.
+     * @param objectClasses set of RDF objects
+     */
+    void setObjectClasses(Set<RDF> objectClasses);
```
Comment (acoburn) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: Ditto.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java (hunk 5 above)
Comment (ajs6f) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: +1

Comment (ajs6f) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: If this kind of action is really desperately desired, a fluent builder apparatus is the way to go.

File: src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+     * @return set of FedoraResource
+     */
+    Set<FedoraResource> getObjects();
+
+    /**
+     * Get the set of RDF describing the rdf:types for this ACL, empty set if none.
+     *
+     * @return set of RDF objects
+     */
+    Set<RDF> getObjectClasses();
```
Comment (peichman-umd) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: Shouldn't the types be a `Set<URI>`?

Comment (whikloj) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: My understand is that the object classes are the values for the rdf:type which are (I believe) already mapped to RDF objects from the jena vocabulary. But I can take it down to a URI.

Comment (peichman-umd) [inline src/main/java/org/fcrepo/auth/webac/WebACAuthorization.java]: @whikloj The com.hp.hpl.jena.vocabulary.RDF class is (AFAICT) a static constants class that returns Jena Property or Resource objects for items in the RDF vocabulary (e.g., rdf:value, rdf:Seq). We could certainly use `RDF.type` from that class in a implementation to locate the type triples, but the returned classes would either be `URI` or ` com.hp.hpl.jena.rdf.model.Resource` objects. My vote is for `URI`, to keep from tying us at this level to Jena.

**Comment by peichman-umd** [comment]
:+1:

**Comment by acoburn** [comment]
:+1:

**Comment by whikloj** [comment]
I agree, what fine work. Someone press the button.

**Comment by awoods** [comment]
Someone should squash the commits before pressing the button.

**Comment by whikloj** [comment]
I'll give it a try, never done that before.

**Comment by awoods** [comment]
i.e.
>git checkout --track origin pull/4
>git log
>git rebase -i <commit-id before first commit>
>> "pick" first, "s" others
>> update comment
>git push origin pull/4:master
----------------------
Assuming you already did a "git pull" with the following ".git/config" update:
```bash
[remote "origin"]
        url = https://github.com/fcrepo4-labs/fcrepo-module-auth-webac.git
        fetch = +refs/heads/*:refs/remotes/origin/*
        fetch = +refs/pull/*/head:refs/remotes/origin/pull/*
```

**Comment by whikloj** [comment]
I hope that works.

**Comment by awoods** [comment]
I does not look like it. Now I see 7 commits instead of one squashed commit. Was that intentional, @whikloj?

**Comment by whikloj** [comment]
Sorry, I should have asked. I just removed some of the commits. I'll squash it all.

**Pull request merged**